packages/signers/src/add-signers.ts

```diff
@@ -0,0 +1,42 @@
+import { IInstruction, isSignerRole } from '@solana/instructions';
+import { BaseTransaction } from '@solana/transactions';
+
+import { IAccountSignerMeta, IInstructionWithSigners, ITransactionWithSigners } from './account-signer-meta';
+import { TransactionSigner } from './transaction-signer';
+
+/** Attaches the provided signers to the account metas of an instruction when applicable. */
+export function addSignersToInstruction<TInstruction extends IInstruction>(
+    signers: TransactionSigner[],
+    instruction: TInstruction | (TInstruction & IInstructionWithSigners)
+): TInstruction & IInstructionWithSigners {
+    if (!instruction.accounts || instruction.accounts.length === 0) {
+        return instruction as TInstruction & IInstructionWithSigners;
+    }
+
+    const signerByAddress = new Map(signers.map(signer => [signer.address, signer]));
```

Comment:
What should we do if someone hands you >1 signer for the same address? This code will just choose the last one and drop the rest.

Reply:
How about we plug in the `deduplicateSigners` logic in there? That way it will deduplicate by reference but fail if distinct signers are provided for the same address.

Reply:
I updated the code and added a test for it. ✅